### PR TITLE
feat: assign default admin role to deployer

### LINKED ISSUE #10: Governor Bravo proposals to change variables in NounsAuctionHouse.sol
Referring to branch [`update-duration`, ](https://github.com/bostondao/nouns-monorepo-fork/pull/7/files) the methods to change auction variables are only callable by the owner address. We are looking to modify the following methods to enable GB proposals to execute after a vote has passed:

- setTimeBuffer
- setReservePrice
- setMinBidIncrementPercentage
- setDuration

Questions: Can a Governor Bravo passed vote execute a method from the `owner` address (multi-sig) that deployed the contracts?

Also open to other ideas to make these variables configurable after a successful proposal passes.

Thanks in advance!

### PR DESCRIPTION
**Pull request by jackdishman**

### CODE REVIEW

**Comment by colink14110476** [comment]
The issue with the error popping up is ContextUpgradable is inherited which defines _msgData and _msgSender:
Its defined as such
```
    function _msgSender() internal view virtual returns (address) {
        return msg.sender;
    }

    function _msgData() internal view virtual returns (bytes calldata) {
        return msg.data;
    }
```
Now `AccessControl` inherits Context which defines both of those. 

To work around this you need to override those two functions for the two classes like so:
```
    function _msgSender() internal view override(Context, ContextUpgradeable) returns (address) {
        return msg.sender;
    }

    function _msgData() internal view override(Context, ContextUpgradeable) returns (bytes calldata) {
        return msg.data;
    }
}
```

Alternatively and the better approach is to use AccessControlUpgradeable instead of AccessControl. These are essentially identical, but AccessControlUpgradeable inherits ContextUpgradeabl instead of Context. Ie this should work:
```
diff --git a/packages/nouns-contracts/contracts/NounsAuctionHouse.sol b/packages/nouns-contracts/contracts/NounsAuctionHouse.sol
index 4f56366e..50bca724 100644
--- a/packages/nouns-contracts/contracts/NounsAuctionHouse.sol
+++ b/packages/nouns-contracts/contracts/NounsAuctionHouse.sol
@@ -24,8 +24,7 @@
 
 pragma solidity ^0.8.6;
 
-import "@openzeppelin/contracts/access/AccessControl.sol";
-
+import { AccessControlUpgradeable } from '@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol';
 import { PausableUpgradeable } from '@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol';
 import { ReentrancyGuardUpgradeable } from '@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol';
 import { OwnableUpgradeable } from '@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol';
@@ -34,7 +33,7 @@ import { INounsAuctionHouse } from './interfaces/INounsAuctionHouse.sol';
 import { INounsToken } from './interfaces/INounsToken.sol';
 import { IWETH } from './interfaces/IWETH.sol';
 
-contract NounsAuctionHouse is INounsAuctionHouse, PausableUpgradeable, ReentrancyGuardUpgradeable, OwnableUpgradeable, AccessControl {
+contract NounsAuctionHouse is INounsAuctionHouse, PausableUpgradeable, ReentrancyGuardUpgradeable, OwnableUpgradeable, AccessControlUpgradeable {
     // The Nouns ERC721 token contract
     INounsToken public nouns;
 
```

**Comment by jackdishman** [comment]
The GB address(es) will need to be given the `DEFAULT_ADMIN_ROLE`, right?

**Comment by colink14110476** [comment]
Yup, the best way would be to pass the `timelock`/`executor` address as a parameter into the initializer function and set to the admin role. 
Since we inherit from accessControl, we also have the functions grantRole available to use whenever.

**Comment by jackdishman** [comment]
Ok I'll modify the **NounsAuctionHouse** to add two parameters for `executor` address on the _initializer_ function.

Also, I got some feedback from Rick via Telegram that we should use a `bravoOrOwner` modifier instead of `DEFAULT_ADMIN_ROLE ` in case there are other admin functions we don't want voting to control. Thoughts on this @colink14110476 ?

**Comment by colink14110476** [comment]
> Ok I'll modify the **NounsAuctionHouse** to add two parameters for `executor` address on the _initializer_ function.
> 
> Also, I got some feedback from Rick via Telegram that we should use a `bravoOrOwner` modifier instead of `DEFAULT_ADMIN_ROLE ` in case there are other admin functions we don't want voting to control. Thoughts on this @colink14110476 ?

I'm cool with either, one benefit with the DEFAULT_ADMIN_ROLE is that its not as limiting as `bravoOrOwner` ie its easy to revoke/add admins. For immutability, the bravoOrOwner may be preferable.

**Comment by jackdishman** [comment]
I created a new role because I believe the idea is that we want it to not be easy to add/revoke admins and it should only be the two addresses

**Comment by colink14110476** [comment]
> I created a new role because I believe the idea is that we want it to not be easy to add/revoke admins and it should only be the two addresses

Since we don't want to use the add/revoke admin stuff, the accessControl may now be overkill. We could just have a modifier function, this would save a good amount of gas since we don't need to store the admin values:
```
    modifier onlyBravoOrOwner() {
        checkBravorOrOwner ();
        _;
    }
    
    function checkBravorOrOwner() internal view virtual {
        require(msg.sender == owner() || msg.sender == executorAddress, "Caller is not bravo or owner");
    }
```

**Comment by jackdishman** [comment]
For this modification,
```
    modifier onlyBravoOrOwner() {
        checkBravoOrOwner ();
        _;
    }
    
    function checkBravoOrOwner() internal view virtual {
        require(msg.sender == owner() || msg.sender == executorAddress, "Caller is not bravo or owner");
    }
```
should executorAddress in `checkBravoOrOwner` method be passed in as a param to the initialize function or how do you recommend defining this?

Then I'll replace these lines:
`require(hasRole(BRAVO_OR_OWNER, msg.sender), "Caller is not an admin");`
with:
`checkBravoOrOwner();`

**Comment by colink14110476** [comment]
> For this modification,
> 
> ```
>     modifier onlyBravoOrOwner() {
>         checkBravoOrOwner ();
>         _;
>     }
>     
>     function checkBravoOrOwner() internal view virtual {
>         require(msg.sender == owner() || msg.sender == executorAddress, "Caller is not bravo or owner");
>     }
> ```
> 
> should executorAddress in `checkBravoOrOwner` method be passed in as a param to the initialize function or how do you recommend defining this?
> 
> Then I'll replace these lines: `require(hasRole(BRAVO_OR_OWNER, msg.sender), "Caller is not an admin");` with: `checkBravoOrOwner();`

Yup the executorAddress can be passed in and set in the intialize function.
I would recommend creating a new field:
```
address public immutable executorAddress;
```
As for replacing lines, you can rm the `hasRole` lines and replace when `onlyOwner` is used with `onlyBravoOrOwner`.

**Comment by jackdishman** [comment]
> `address public immutable executorAddress;`
Since there's no constructor (only an _initalize_ method), it's not letting me set the immutable `executorAddress` value. Should I just use constant instead?

**Comment by colink14110476** [comment]
> > `address public immutable executorAddress;`
> > Since there's no constructor (only an _initalize_ method), it's not letting me set the immutable `executorAddress` value. Should I just use constant instead?

ah woops my bad, neither immutable nor constant will work actually, since the immutable/constant require it being set at the constructor level. just `address public executorAddress;` without the immutable/constant should be fine.

File: packages/nouns-contracts/contracts/NounsAuctionHouse.sol [hunk 1]
```diff
@@ -51,6 +51,12 @@ contract NounsAuctionHouse is INounsAuctionHouse, PausableUpgradeable, Reentranc
     // The duration of a single auction
     uint256 public duration;
 
+    // The role to modify contract variables
+    bytes32 public constant BRAVO_OR_OWNER = keccak256("bravoOrOwner");
```
Comment (colink14110476) [inline packages/nouns-contracts/contracts/NounsAuctionHouse.sol]: rm

**Comment by colink14110476** [comment]
code changes lgtm. A merge is needed to resolve conflicts. this can be done via `git merge origin/master` 
Test file, `auction.test.ts` needs to be updated as well. Passing any address into the new argument on ln 35 should work.

**Comment by jackdishman** [comment]
> Test file, auction.test.ts needs to be updated as well. Passing any address into the new argument on ln 35 should work

I passed the executor address into the deploy method test in the latest commit and also into the `initialize` tx

Merge conflicts resolved as well!

**Comment by colink14110476** [comment]
test cases are currently failing:
these can be verified via `npx hardhat test`

File: packages/nouns-contracts/test/auction.test.ts [hunk 2]
```diff
@@ -39,6 +41,7 @@ describe('NounsAuctionHouse', () => {
       RESERVE_PRICE,
       MIN_INCREMENT_BID_PERCENTAGE,
       DURATION,
+      executor,
```
Comment (colink14110476) [inline packages/nouns-contracts/test/auction.test.ts]: this shsould be executor.address

File: packages/nouns-contracts/test/auction.test.ts [hunk 3]
```diff
@@ -47,7 +50,7 @@ describe('NounsAuctionHouse', () => {
 
     nounsToken = await deployNounsToken(deployer, noundersDAO.address, deployer.address);
     weth = await deployWeth(deployer);
-    nounsAuctionHouse = await deploy(deployer);
+    nounsAuctionHouse = await deploy(deployer, EXECUTOR);
```
Comment (colink14110476) [inline packages/nouns-contracts/test/auction.test.ts]: EXECUTOR is a string so the wrong type is being passed in.

**Comment by colink14110476** [comment]
end2end.ts also needs to update its deployment of the nounsAuctionHouse

File: packages/nouns-contracts/test/auction.test.ts [hunk 4]
```diff
@@ -325,4 +329,8 @@ describe('NounsAuctionHouse', () => {
       .to.emit(nounsAuctionHouse, 'AuctionSettled')
       .withArgs(nounId, '0x0000000000000000000000000000000000000000', 0);
   });
+
+  it('should set the duration of an auction', async () => {
+    await (await nounsAuctionHouse.setDuration([60 * 60 * 25])) // sets duration to 25 hours
```
Comment (colink14110476) [inline packages/nouns-contracts/test/auction.test.ts]: incorrect type, pass 60 * 60 * 25 not [...]

Comment (colink14110476) [reply packages/nouns-contracts/test/auction.test.ts]: You should also write some `expect` statements

**Comment by colink14110476** [comment]
This is my recommended diff to fix the existing test issues:
```
diff --git a/packages/nouns-contracts/test/auction.test.ts b/packages/nouns-contracts/test/auction.test.ts
index cb5c81c6..12a9a65b 100644
--- a/packages/nouns-contracts/test/auction.test.ts
+++ b/packages/nouns-contracts/test/auction.test.ts
@@ -23,6 +23,7 @@ describe('NounsAuctionHouse', () => {
   let noundersDAO: SignerWithAddress;
   let bidderA: SignerWithAddress;
   let bidderB: SignerWithAddress;
+  let executor: SignerWithAddress;
   let snapshotId: number;
 
   const TIME_BUFFER = 15 * 60;
@@ -30,9 +31,8 @@ describe('NounsAuctionHouse', () => {
   const MIN_INCREMENT_BID_PERCENTAGE = 5;
   const DURATION = 60 * 60 * 24;
   // Address of Govenor Bravo executor
-  const EXECUTOR = `0x308112D06027Cd838627b94dDFC16ea6B4D90004`;
 
-  async function deploy(deployer?: SignerWithAddress, executor?: SignerWithAddress) {
+  async function deploy(deployer?: SignerWithAddress) {
     const auctionHouseFactory = await ethers.getContractFactory('NounsAuctionHouse', deployer);
     return upgrades.deployProxy(auctionHouseFactory, [
       nounsToken.address,
@@ -41,16 +41,15 @@ describe('NounsAuctionHouse', () => {
       RESERVE_PRICE,
       MIN_INCREMENT_BID_PERCENTAGE,
       DURATION,
-      executor,
+      executor.address
     ]) as Promise<NounsAuctionHouse>;
   }
 
   before(async () => {
-    [deployer, noundersDAO, bidderA, bidderB] = await ethers.getSigners();
-
+    [deployer, noundersDAO, bidderA, bidderB, executor] = await ethers.getSigners();
     nounsToken = await deployNounsToken(deployer, noundersDAO.address, deployer.address);
     weth = await deployWeth(deployer);
-    nounsAuctionHouse = await deploy(deployer, EXECUTOR);
+    nounsAuctionHouse = await deploy(deployer);
 
     const descriptor = await nounsToken.descriptor();
 
@@ -75,7 +74,7 @@ describe('NounsAuctionHouse', () => {
       RESERVE_PRICE,
       MIN_INCREMENT_BID_PERCENTAGE,
       DURATION,
-      EXECUTOR,
+      executor.address,
     );
     await expect(tx).to.be.revertedWith('Initializable: contract is already initialized');
   });
@@ -331,6 +330,13 @@ describe('NounsAuctionHouse', () => {
   });
 
   it('should set the duration of an auction', async () => {
-    await (await nounsAuctionHouse.setDuration([60 * 60 * 25])) // sets duration to 25 hours
-  })
+    await (await nounsAuctionHouse.unpause()).wait();
+    await (await nounsAuctionHouse.setDuration(60 * 60 * 25)) // sets duration to 25 hours
+    let newDuration = await nounsAuctionHouse.duration();
+    await expect(newDuration).to.be.equal(60 * 60 * 25)
+
+    await nounsAuctionHouse.connect(executor).setDuration(60 * 60 * 26) // sets duration to 26 hours
+    newDuration = await nounsAuctionHouse.duration();
+    await expect(newDuration).to.be.equal(60 * 60 * 26)
+  });
 });
diff --git a/packages/nouns-contracts/test/end2end.test.ts b/packages/nouns-contracts/test/end2end.test.ts
index 7385e69a..4de6d588 100644
--- a/packages/nouns-contracts/test/end2end.test.ts
+++ b/packages/nouns-contracts/test/end2end.test.ts
@@ -100,6 +100,7 @@ async function deploy() {
     RESERVE_PRICE,
     MIN_INCREMENT_BID_PERCENTAGE,
     DURATION,
+    deployer.address
   ]);
 
   // 2b. CAST proxy as AuctionHouse
```
[diff.txt](https://github.com/bostondao/nouns-monorepo-fork/files/10311630/diff.txt)

You could apply this with `git apply diff.txt`

**Review by colink14110476 — APPROVED** [review]

**Comment by colink14110476** [comment]
pause and unpause should also be onlyOwnerOrBravo

**Pull request merged**